model/renderNativeWindow.cc

```diff
@@ -54,41 +63,127 @@ using namespace ecolab;
+    return childSurface;
   }
 #endif
-  
 
-  void RenderNativeWindow::renderToNativeWindow(unsigned long window)
+  WindowInformation::WindowInformation()
+  {
+  }
+  WindowInformation::~WindowInformation()
+  {
+    cout << "Delete called for window information" << endl;
+  }
+
+  void WindowInformation::initialize(unsigned long parentWin, int left, int top, int cWidth, int cHeight)
+  {
+    parentWindowId = parentWin;
+    offsetLeft = left;
+    offsetTop = top;
+
+    static bool errorHandlingSet = (XSetErrorHandler(throwOnXError), true);
+    display = XOpenDisplay(nullptr);
+    int err = XGetWindowAttributes(display, parentWin, &wAttr);
+    if (err > 1)
+      throw runtime_error("Invalid window: " + to_string(parentWin));
+
+    childWidth = wAttr.width - offsetLeft;
+    childHeight = wAttr.height - offsetTop;
+    
+    // Todo:: currently width, height parameters are not getting passed properly
+    // Eventually, we need those in order to ensure we don't need to worry about
+    // paddings, scrollbars etc
+    if(cWidth > 0) {
+      childWidth = min(childWidth, cWidth);
+    }
+    if(cHeight > 0) {
+      childHeight = min(childHeight, cHeight);
+    }
+
+    cout << childWidth << "::" << childHeight << "::" << offsetLeft << "::" << offsetTop << endl;
+    childWindowId = XCreateSimpleWindow(display, parentWin, offsetLeft, offsetTop, childWidth, childHeight, 0, 0, 0); //TODO:: Should we pass visual and attributes at the end?
+    XMapWindow(display, childWindowId);
+  }
+
+  void WindowInformation::copy(WindowInformation *winInfo)
+  {
+    this->parentWindowId = winInfo->parentWindowId;
+    this->childWindowId = winInfo->childWindowId;
+    this->offsetLeft = winInfo->offsetLeft;
+    this->offsetTop = winInfo->offsetTop;
+    this->display = winInfo->display;
+    this->wAttr = winInfo->wAttr;
+    // this->childSurface = winInfo->childSurface;// TODO:: Later, try reusing the surface also in winInfo
+  }
+
+  RenderNativeWindow::RenderNativeWindow()
+  {
+    winInfo = new WindowInformation(); // TODO:: Try to reuse the WindowInformation object instead of creating a new one.... perhaps it should be a static data member?
+  }
+
+  RenderNativeWindow::~RenderNativeWindow()
+  {
+    delete winInfo;
+  }
+
+  RenderNativeWindow &RenderNativeWindow::operator=(const RenderNativeWindow &a)
+  {
+    cout << "Calling assignment on renderNativeWindow" << endl;
+    a.winInfo->copy(this->winInfo);
+    return *this;
+  }
+
+  RenderNativeWindow::RenderNativeWindow(const RenderNativeWindow &a)
   {
-    auto tmp=nativeWindowSurface(window);
+    cout << "Copy constructor for RenderNativeWindow called... TODO:: implement this" << endl;
+    //a.winInfo->copy(this->winInfo); // reverse?
+  };
+
+  void RenderNativeWindow::renderFrame()
+  {
+    auto tmp = createNativeWindowSurface(winInfo);
+
+    //TODO:: Review if this paint (below 3 lines) is really needed with each frame
```

Comment:
My intuition is that painting the background canvas is not necessary for X11, but might be for Windows. I guess we should suck it and see.